### PR TITLE
first steps in building sequential docking fn

### PR DESCRIPTION
**Pull request by JenkeScheen**
## Description
When debugging the multiprocessing function in `run_docking_oe.py` it becomes unnecessarily complex to figure out what's going wrong in daughter processes. It's more efficient to be able to check if a non-multiprocessed job is failing (and to see direct error logs); this PR implements that functionality.

We'll set `-n 0` to be the flag to disable multiprocessing.

## Todos
  - [ ] write base function and debug
  - [ ] test on use-case

## Status
- [ ] Ready to go

### CODE REVIEW

**Comment by JenkeScheen** [comment]
The use-case is somewhat open-ended, all I know is docking is failing at the moment for large (>50 compounds) jobs, resulting in `Bus error (core dumped)` (on interactive node) and ~8GB core dumps in working directory. 

During debugging I'm finding that the core is dumped while loading design units into memory. Currently there are 576 Mpro DUs being loaded; core is dumped consistently at the 506th:
```
 88%|█████████████████████████████████████████████████████████▌               | 506/576 

[00:51<00:06, 11.55it/s]

/lila/data/chodera/asap-datasets/current/sars_01_prepped_v2/Mpro-x3080_0A_GAB-REV-70cc3ca5-8/prepped_receptor_0.oedu
```
--> temporarily removed the file, then core is just dumped at another, this time the 505th. Could it just be a memory issue?

**Comment by JenkeScheen** [comment]
right, so truncating the input to only loading in 500 DUs passes that stage. Then, weirdly, the process is just 'killed'?
```
Loading 500 design units..
100%|████████████████| 500/500 [01:39<00:00,  5.03it/s]
Killed
```

**Comment by JenkeScheen** [comment]
@kaminow @apayne97 this is WIP, lmk if you have any ideas

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 1]
```diff
@@ -404,9 +479,14 @@ def main():
...
-    with mp.Pool(processes=nprocs) as pool:
-        results_df = pool.starmap(mp_func, mp_args)
+    if nprocs == 0:
+        print(f"Running {len(mp_args)} docking runs sequentially.")
+        results_df = dock_sequentially(mp_args)
+
+    else:
+        print(f"Running {len(mp_args)} docking runs over {nprocs} cores.")
+        with mp.Pool(processes=nprocs) as pool:
+            results_df = pool.starmap(mp_func, mp_args)
```
Comment (kaminow) [inline asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:489]: I may be missing something, but I think we can just use `mp_func` in the non-multiprocessed version as well, eg
```suggestion
    if nprocs == 0:
        print(f"Running {len(mp_args)} docking runs sequentially.")
        results_df = [mp_func(*a) for a in mp_args]
    else:
        print(f"Running {len(mp_args)} docking runs over {nprocs} cores.")
        with mp.Pool(processes=nprocs) as pool:
            results_df = pool.starmap(mp_func, mp_args)
```

and then get rid of the `dock_sequentially` function

Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:489]: I agree if it is a one for one match, but lets rename it to `dock_something` to make it clear what the function is doing. `mp_func ` is not clear.

**Review by hmacdope — CHANGES_REQUESTED** [review]

**Comment by apayne97** [comment]
@JenkeScheen I think this is happening because we're using more of the structures now, I think it just requires more data than before. We could change the script to somehow batch load the jobs, but at least in small scale just changing the amount of requested memory seems to fix the problem for me:

```
2023-03-28 15:25:56,340 | INFO | Using /lila/data/chodera/asap-datasets/current/sars_01_prepped_v3/*/*_prepped_receptor_0.oedu as glob
2023-03-28 15:25:56,601 | INFO | Loading 576 design units
2023-03-28 15:26:53,637 | INFO | 576 design units loaded
2023-03-28 15:26:53,639 | INFO | 220 molecules found
2023-03-28 15:26:53,639 | INFO | 576 receptor structures found
2023-03-28 15:26:53,640 | INFO | Using Compound_ID column from sdf file
2023-03-28 15:26:53,925 | INFO | CPUs: 128
2023-03-28 15:26:53,925 | INFO | N Processes: 126720
2023-03-28 15:26:53,925 | INFO | N Cores: 32
2023-03-28 15:26:53,937 | INFO | Running 32 docking runs over 32 cores.
```

**Comment by apayne97** [comment]
@JenkeScheen what's your submission script?

**Comment by JenkeScheen** [comment]
my above issues have been resolved though - turns out it was just a problem with the formatting of the SDF (occasional Windows newline characters - '^M' - were throwing wrenches in the works). Closing this PR for now as @hmacdope and I will write a docking CLI specific for ASAP project support at some point soon.

**Comment by apayne97** [comment]
> @apayne97 this was on an interactive node, i.e. with 1 core with 4GB. I did run the docking code with `-n 1` so *think* that should have been okay.
> 
> 
> 
> 

Was this with the current amount of DUs or from before?

**Comment by JenkeScheen** [comment]
current amount, so `n=576`

**Pull request closed without merging**